Move queries to exandra, update configurations and remove deprecated options and dependencies

eddbbt opened the pull request:
Update select queries from cqex to exandra (following recent updates in astarte)
Update connection config to include xandra_nodes alongside cassandra_nodes 
Update xandra_options management 
Remove deprecated autodiscovery options for xandra (after 0.15.0 it's always active)
Using now astarte_core github release branch instead of a release (latest update still not released)
Edit test with a workaround for exandra (tests still expect :astarte_data_access_xandra, that does nto exist anymore, this workaround register the cild again only for testing purposes)
Remove xandra_utils module, as those are not used anymore and deprecated in astarte context (breaking)
Register again cqex and cqerl, as it's too early to remove them

Suggestions for file path and/or naming changes are welcome

Annopaolo commented on lib/astarte_data_access/device/deletion_in_progress.ex:
```diff
@@ -0,0 +1,36 @@
+  typed_schema "deletion_in_progress" do
+    field(:vmq_ack, :boolean)
+    field(:dup_start_ack, :boolean)
+    field(:dup_end_ack, :boolean)
```
If you add `:ecto` to `import_deps` in `formatter.exs` you will have back the synatx without parentheses

Annopaolo replied:
That addition is in another commit, but I might have missed the syntax change in schema files

Annopaolo commented on lib/astarte_data_access/devices/device.ex:
I suggest having this file outside of the directory, like e.g. `kv_store.ex`

Annopaolo commented on lib/astarte_data_access/groups/group.ex:
```diff
@@ -0,0 +1,46 @@
+#
+# This file is part of Astarte.
+#
+# Copyright 2019 Ispirata Srl
```
Since the schema files are taken from a previous effort made in 2025, I would expect copyright to be coherent. If not, it is ok to have another commit with mass update to copyright, but seems sus

Annopaolo commented on lib/astarte_data_access/kv_store.ex:
```diff
@@ -0,0 +1,96 @@
+# limitations under the License.
+#
+# TODO: Copied from astarte_data_access PR #71, see: https://github.com/astarte-platform/astarte_data_access/pull/71
+# use `astarte_data_access` when it will be merged
```
Leftover?

Annopaolo commented on mix.exs:
```diff
@@ -64,17 +63,22 @@ defmodule Astarte.DataAccess.Mixfile do
-      {:astarte_core, "~> 1.2"}
+      {
+        :astarte_core,
+        github: "noaccOS/astarte_core", branch: "backport/refactor-standard-ecto-functions"
```
Do not merge until this is changed

Annopaolo commented on lib/astarte_data_access/database.ex:
```diff
@@ -43,7 +43,7 @@ defmodule Astarte.DataAccess.Database do
 
       any_error ->
-        Logger.warn("Failed connection to the database. Reason: #{inspect(any_error)}")
+        Logger.warning("Failed connection to the database. Reason: #{inspect(any_error)}")
```
Let's resolve warning like that in a different PR (or commit, at least)

Annopaolo commented on lib/astarte_data_access/xandra_utils.ex:
Are we sure to remove this file? Dependents on `astarte_data_access` that use e.g. `run` will break

noaccOS replied:
They will break anyway if we don't change it to use Exandra's `Xandra.Cluster`, as DataAccess' was removed from the supervision tree

eddbbt committed:
Fix test for Exandra

Use strings instead of atoms as keys in maps
Add missing fields to mocked devices table
Apply workaround for register :astarte_data_access_xandra child even without the actual :astarte_data_access_xandra process (pointing to DataAccess.Repo)
This needs to be removed with a complete test rewrite

Signed-off-by: Eddy Babetto <eddy.babetto@secomind.com>

eddbbt committed:
Update Astarte Core dependency

UPdate astarte core dependency as it now includes required functionality for next edits in data_access

Signed-off-by: Eddy Babetto <eddy.babetto@secomind.com>

eddbbt committed:
Update queries to Exandra

Signed-off-by: Eddy Babetto <eddy.babetto@secomind.com>

eddbbt committed:
Update Copyright info

Signed-off-by: Eddy Babetto <eddy.babetto@secomind.com>

eddbbt committed:
Update configuration to work with Exandra

Removed Xandra.Cluster from supervision tree
Removed deprecated options (autodiscovery)

Signed-off-by: Eddy Babetto <eddy.babetto@secomind.com>

eddbbt committed:
Remove Xandra_utils module, as it was not used anymore in astarte

Signed-off-by: Eddy Babetto <eddy.babetto@secomind.com>

eddbbt committed:
Update changelog

Signed-off-by: Eddy Babetto <eddy.babetto@secomind.com>

eddbbt committed:
fix fetch_value returns value on KvStore

Signed-off-by: Eddy Babetto <eddy.babetto@secomind.com>

lusergit commented on lib/astarte_data_access/kv_store.ex:
```diff
@@ -89,6 +89,10 @@ defmodule Astarte.DataAccess.KvStore do
-    Repo.fetch_by(query, primary_key, opts)
+    case Repo.fetch_by(query, primary_key, opts) do
+      nil -> {:error, nil}
+      row -> {:ok, row}
+      {:error, error} -> {:error, error}
```
the CI is already suggesting this but what I think you want to do here is
```suggestion
      {:error, error} -> {:error, error}
      row -> {:ok, row}
```
as otherwise the first row always matches

lusergit commented on lib/astarte_data_access/interface.ex:
```diff
@@ -17,83 +17,68 @@
-        [%{count: 0}] ->
-          {:error, :interface_not_found}
-      end
+    case Repo.aggregate(query, :count) do
+      1 -> :ok
+      0 -> {:error, :interface_not_found}
```
Maybe
```suggestion
    case Repo.some?(query) do
      {:ok, true} -> :ok
      {:ok, false} -> {:error, :interface_not_found}
```
Is more clear but I don't have a strong opinion about it

lusergit reviewed (changes requested):
Overall lgtm :+1:  small nits

eddbbt committed:
Add dateTime format

As per astarte PR #1134

Signed-off-by: Eddy Babetto <eddy.babetto@secomind.com>

Annopaolo commented on lib/astarte_data_access/database.ex:
This file has not been touched, no copyright update is needed

Annopaolo commented on mix.lock:
```diff
@@ -1,13 +1,13 @@
+  "astarte_core": {:git, "https://github.com/astarte-platform/astarte_core.git", "7ba3d8f672e54c55b0abe8fd7c4d00f40a4f023e", [branch: "release-1.2"]},
   "castore": {:hex, :castore, "1.0.12", "053f0e32700cbec356280c0e835df425a3be4bc1e0627b714330ad9d0f05497f", [:mix], [], "hexpm", "3dca286b2186055ba0c9449b4e95b97bf1b57b47c1f2644555879e659960c224"},
-  "cqerl": {:hex, :cqerl, "2.1.3", "bab07ea05fabf524112eaf199c349bbb05bc7cc2ac3655db117650ff86c963ec", [:rebar3], [{:lz4, "~> 0.2.4", [hex: :lz4_erl, repo: "hexpm", optional: false]}, {:re2, "1.9.8", [hex: :re2, repo: "hexpm", optional: false]}, {:semver, "~> 0.0.1", [hex: :semver_erl, repo: "hexpm", optional: false]}, {:snappyer, "1.2.8", [hex: :snappyer, repo: "hexpm", optional: false]}, {:uuid ...
+  "cqerl": {:hex, :cqerl, "2.0.1", "c92929e885adc50cda1f11b73eb0121363e8418533312f8a44defb5f14853445", [:rebar3], [{:lz4, "~> 0.2.4", [hex: :lz4_erl, repo: "hexpm", optional: false]}, {:re2, "1.9.5", [hex: :re2, repo: "hexpm", optional: false]}, {:semver, "~> 0.0.1", [hex: :semver_erl, repo: "hexpm", optional: false]}, {:snappyer, "1.2.6", [hex: :snappyer, repo: "hexpm", optional: false]}, {:uuid ...
```
Let's not downgrade a dependency we (still) need

Annopaolo commented on mix.lock:
```diff
@@ -25,12 +25,13 @@
   "protobuf": {:hex, :protobuf, "0.14.1", "9ac0582170df27669ccb2ef6cb0a3d55020d58896edbba330f20d0748881530a", [:mix], [{:jason, "~> 1.2", [hex: :jason, repo: "hexpm", optional: true]}], "hexpm", "39a9d49d346e3ed597e5ae3168a43d9603870fc159419617f584cdf6071f0e25"},
   "quickrand": {:hex, :quickrand, "2.0.7", "d2bd76676a446e6a058d678444b7fda1387b813710d1af6d6e29bb92186c8820", [:rebar3], [], "hexpm", "b8acbf89a224bc217c3070ca8bebc6eb236dbe7f9767993b274084ea044d35f0"},
-  "re2": {:hex, :re2, "1.9.8", "d125d43b1d695d35f3625f11d984e4417192240a8d0bec89b1d6872b8f861d7b", [:rebar3], [], "hexpm", "3aa83a5c50f88cf1578642020e1e63918053911882045e4981e489e3dd74b7e8"},
+  "re2": {:hex, :re2, "1.9.5", "3c419527fb2cc75eda1657f04dc7e8cea9864899b43ff6cc3250fa7525431e83", [:rebar3], [], "hexpm", "4861336271ac565224e79e133cbaf90af77739cda3fff25f6965b24a1146a4d6"},
```
See above, this will break on ARM

Annopaolo commented on CHANGELOG.md:
```diff
@@ -4,6 +4,13 @@ All notable changes to this project will be documented in this file.
+### Changed
+- BREAKING: Removed XandraUtils module because it's not 
+  needed anymore inside the main Astarte project
+- removed CQerl (unmantained) and moved queries to Ecto + Xandra
```
```suggestion
- Changed the database driver from CQEx (unmantained) to Exandra
```

eddbbt committed:
Update Repo file

Update fetch_one repo function and all its uses

Signed-off-by: Eddy Babetto <eddy.babetto@secomind.com>
